man/tpm2_certifycreation.1.md
% tpm2_certifycreation(1) tpm2-tools | General Commands Manual

# NAME

**tpm2_certifycreation**(1) - Attest the association between a loaded public
area and the provided hash of the creation data.

# SYNOPSIS

**tpm2_certifycreation** [*OPTIONS*]

# DESCRIPTION

**tpm2_certifycreation**(1) - Attest the association between a loaded public
area and the provided hash of the creation data. The creation data and the
creation ticket is produced when creating the object. The object itself is
created with either **TPM2_CreatePrimary** or **TPM2_Create** commands.

# OPTIONS

  * **-C**, **\--signingkey-context**=_OBJECT_:

    Context object pointing to the key used that signs the attestation.

  * **-P**, **\--signingkey-auth**_AUTH_:

    Optional authorization value to use for the key specified by **-C**.

  * **-c**, **\--certifiedkey-context**=_OBJECT_:

    Context object pointing to the key that has to be certified.

  * **-g**, **\--hash-algorithm**=_ALGORITHM_:

    The hash algorithm used to digest the creation data.

  * **-s**, **\--scheme**=_ALGORITHM_:

    The signing scheme used to sign the attestation data.

  * **-d**, **\--creation-hash**=_FILE_

    File containing the digest of the creation data.

  * **-t**, **\--ticket**=_FILE_:

    The ticket file to validate that the creation data was produced by the TPM.

  * **-o**, **\--signature**=_FILE_:

    File containing the signature of the attestation data for the certified key.

  * **-f**, **\--format**=_FORMAT_:

    Output signature format selection.

  * **--attestation**=_FILE_:

    The attestation data of the type TPM2_CREATION_INFO signed with signing key.

  * **-q**, **\--qualification**=_FILE\_OR\_HEX_:

    Optional, the policy qualifier data that the signer can choose to include in the
    signature. Can either be a path or hex string.

  * **\--cphash**=_FILE_

    File path to record the hash of the command parameters. This is commonly
    termed as cpHash. NOTE: When this option is selected, The tool will not
    actually execute the command, it simply returns a cpHash, unless rphash is also required.

  * **\--rphash**=_FILE_

    File path to record the hash of the response parameters. This is commonly
    termed as rpHash.

  * **-S**, **\--session**=_FILE_:

    The session created using **tpm2_startauthsession**. This can be used to
    specify an auxiliary session for auditing and or encryption/decryption of
    the parameters.

## References

[context object format](common/ctxobj.md) details the methods for specifying
_OBJECT_.

[authorization formatting](common/authorizations.md) details the methods for
specifying _AUTH_.

[algorithm specifiers](common/alg.md) details the options for specifying
cryptographic algorithms _ALGORITHM_.

[common options](common/options.md) collection of common options that provide
information many users may expect.

[common tcti options](common/tcti.md) collection of options used to configure
the various known TCTI modules.
[signature format specifiers](common/signature.md)

# EXAMPLES

## Certify creation data of a primary key.

```bash
tpm2_createprimary -C o -c prim.ctx --creation-data create.dat \
-d create.dig -t create.ticket

tpm2_create -G rsa -u rsa.pub -r rsa.priv -C prim.ctx -c signing_key.ctx

tpm2_certifycreation -C signing_key.ctx -c prim.ctx -d create.dig \
-t create.ticket -g sha256 -o sig.nature --attestation attestat.ion -f plain \
-s rsassa
```

[returns](common/returns.md)

[footer](common/footer.md)


CHANGELOG
v2.2.2 - next:
*  fix make distcheck
v2.1.1 - next:
* Potential memory leak fix when tcti/sapi initialization fails.
* tpm2_listpcrs: use TPM2_GetCapability to determine PCRs to read
* listpcrs: remove one redundant call to tpm get cap
* listpcrs: fix for unsupported/disabled alg in -L
* build: use supported comment to suppress GCC7 fallthrough warning
* kdfa: allow to build with OpenSSL 1.1.x

v2.1.0 - 2017-7-24
* Fix readx and writex on multiple EINTR returns.
* Add support for the tabrmd TCTI. This is the new default.
* Change default socket port from 2323 (the old resourcemgr) to 2321 (default
  simulator port).
* Cherry-pick fix for CVE-2017-7524.
* Fix tpm2_listpcr command line option handling.
* Fix tpm2_getmanufec memory issues.

v2.0 - 2017-03-29
* Tracked on the milestone: https://github.com/01org/tpm2.0-tools/milestone/2
* Reworked all the tools to support configurable TCTIs, based on build time
  configuration, one can specify the tcti via the --tcti (-T) option to all
  tools.
* tpm2_getrandom interface made -s a positional argument.
* Numerous bug fixes.

v1.1 - 2016-11-04
Added
* travis ci support.
* Allow for unit tests to be enabled selectively.
* tpm2_rc_decode tool: Decode TPM_RC error codes.
* Android Make file
* tpm2_listpersistent: list all persistent objects
* test scripts for tpm2-tools
* tpm2_nvreadlock
* tpm2_getmanufec: retrieve EC from tpm manufacturer server.
* Copy 'common' and 'sample' code from the TPM2.0-TSS repo.

Modified
* tpm2_takeownership: update option -c to use lockout password to clear.
* tpm2_listpcrs: add options -L and -s, rewrite to increase performance.
* tpm2_quote: added -L option to support selection of multiple banks.
* tpm2_quote: add -q option to get qualifying data.
* configure: Use pkg-config to get info about libcurl and libcrypto.
* configure: Use pkg-config to locate SAPI and TCTI headers / libraries.
* tpm2_x: Add -X option to enable password input in Hex format.
* tpm2_nvdefine: Change -X option to -I.
* tpm2-nvwrite: fix for unable to write 1024B+ data.
* tpm2_getmanufec: Fix base64 encoding.
* tpm2_x: fixed a lot of TPM2B failures caused by wrong initialization.
* tpm2_getmanufec: let configure handle libs.
* tpm2_getmanufec: Convert from dos to unix format.
* build: Check for TSS2 library @ configure time.
* build: Detect required TSS2 and TCTI headers.
* build: Use libtool to build the common library
* build: Install all binaries into sbin.
* build: Build common sources into library.
* build: Move all source files to 'src'.
* Makefile.am: Move all build rules into single Makefile.am.
* everything: Use new TCTI headers and fixup API calls.
* everything: Update source to cope with sapi header cleanup.
* tpm2_activatecredential: Updated to support TCG compatible EK
* tpm2_getpubak: Updated to use TCG compatible EK
* tpm2_getpubek: fix ek creation to follow TCG EK profile spec.

Removed
* Windows related code
* depenedency on the TPM2.0-TSS repo source code

v1.0 - 2015-10-19
* 1.0 release
* 29 tools included


tools/tpm2_certifycreation.c
/* SPDX-License-Identifier: BSD-3-Clause */

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

#include "files.h"
#include "log.h"
#include "tpm2.h"
#include "tpm2_tool.h"
#include "tpm2_alg_util.h"
#include "tpm2_convert.h"
#include "tpm2_options.h"

typedef struct tpm_certifycreation_ctx tpm_certifycreation_ctx;
struct tpm_certifycreation_ctx {
    /*
     * Inputs
     */
    struct {
        const char *ctx_path;
        const char *auth_str;
        tpm2_loaded_object object;
    } signing_key;

    struct {
        const char *ctx_path;
        tpm2_loaded_object object;
    } certified_key;

    char *creation_hash_path;
    TPM2B_DIGEST creation_hash;

    char *creation_ticket_path;
    TPMT_TK_CREATION creation_ticket;

    TPM2B_DATA policy_qualifier;
    const char *policy_qualifier_data;

    TPMI_ALG_HASH halg;
    TPMI_ALG_SIG_SCHEME sig_scheme;
    tpm2_convert_sig_fmt sig_format;
    TPMT_SIG_SCHEME in_scheme;

    /*
     * Outputs
     */
    char *signature_path;
    TPMT_SIGNATURE *signature;

    char *certify_info_path;
    TPM2B_ATTEST *certify_info;

   /*
    * Parameter hashes
    */
    char *cp_hash_path;
    TPM2B_DIGEST cp_hash;
    TPM2B_DIGEST *cphash;
    bool is_command_dispatch;
};

static tpm_certifycreation_ctx ctx = {
        .halg = TPM2_ALG_NULL,
        .sig_scheme = TPM2_ALG_NULL,
        .policy_qualifier = TPM2B_EMPTY_INIT,
};

static tool_rc certifycreation(ESYS_CONTEXT *ectx) {

    /*
     * 1. TPM2_CC_<command> OR Retrieve cpHash
     */
    return tpm2_certifycreation(ectx, &ctx.signing_key.object,
        &ctx.certified_key.object, &ctx.creation_hash, &ctx.in_scheme,
        &ctx.creation_ticket, &ctx.certify_info, &ctx.signature,
        &ctx.policy_qualifier, ctx.cphash);
}

static tool_rc process_output(void) {

    /*
     * 1. Outputs that do not require TPM2_CC_<command> dispatch
     */
    bool is_file_op_success = true;
    if (ctx.cp_hash_path) {
        is_file_op_success = files_save_digest(&ctx.cp_hash, ctx.cp_hash_path);

        if (!is_file_op_success) {
            return tool_rc_general_error;
        }
    }

    if (!ctx.is_command_dispatch) {
        return tool_rc_success;
    }

    /*
     * 2. Outputs generated after TPM2_CC_<command> dispatch
     */
    is_file_op_success = tpm2_convert_sig_save(ctx.signature, ctx.sig_format,
        ctx.signature_path);
    if (!is_file_op_success) {
        LOG_ERR("Failed saving signature data.");
        return tool_rc_general_error;
    }

    is_file_op_success = files_save_bytes_to_file(ctx.certify_info_path,
        ctx.certify_info->attestationData, ctx.certify_info->size);
    if (!is_file_op_success) {
        LOG_ERR("Failed saving attestation data.");
    }

    return is_file_op_success ? tool_rc_success : tool_rc_general_error;
}

static tool_rc process_inputs(ESYS_CONTEXT *ectx) {

    /*
     * 1. Object and auth initializations
     */

    /*
     * 1.a Add the new-auth values to be set for the object.
     */

    /*
     * 1.b Add object names and their auth sessions
     */
    tool_rc rc = tpm2_util_object_load_auth(ectx, ctx.signing_key.ctx_path,
        ctx.signing_key.auth_str, &ctx.signing_key.object, false,
        TPM2_HANDLES_FLAGS_TRANSIENT|TPM2_HANDLES_FLAGS_PERSISTENT);
    if (rc != tool_rc_success) {
        LOG_ERR("Invalid signing key/ authorization.");
        return rc;
    }

    rc = tpm2_util_object_load(ectx, ctx.certified_key.ctx_path,
        &ctx.certified_key.object,
        TPM2_HANDLES_FLAGS_TRANSIENT|TPM2_HANDLES_FLAGS_PERSISTENT);
    if (rc != tool_rc_success) {
        LOG_ERR("Invalid key specified for certification.");
        return rc;
    }

    /*
     * 2. Restore auxiliary sessions
     */

    /*
     * 3. Command specific initializations
     */

     /* Load creation hash */
    rc = files_load_digest(ctx.creation_hash_path, &ctx.creation_hash) ?
        tool_rc_success : tool_rc_general_error;
    if (rc != tool_rc_success) {
        LOG_ERR("Failed loading creation hash.");
        return rc;
    }

    /* Set signature scheme for key type & Validate chosen scheme */
    rc = tpm2_alg_util_get_signature_scheme(ectx,
        ctx.signing_key.object.tr_handle, &ctx.halg, ctx.sig_scheme,
        &ctx.in_scheme);
    if (rc != tool_rc_success) {
        LOG_ERR("bad signature scheme for key type!");
        return rc;
    }

    /* Load creation ticket */
    rc = files_load_creation_ticket(ctx.creation_ticket_path,
        &ctx.creation_ticket) ? tool_rc_success : tool_rc_general_error;
    if (rc != tool_rc_success) {
        LOG_ERR("Could not load creation ticket from file");
        return rc;
    }

    /* Qualifier data is optional. If not specified default to 0 */
    if (ctx.policy_qualifier_data) {
        ctx.policy_qualifier.size = sizeof(ctx.policy_qualifier.buffer);

        rc = tpm2_util_bin_from_hex_or_file(ctx.policy_qualifier_data,
            &ctx.policy_qualifier.size, ctx.policy_qualifier.buffer) ?
            tool_rc_success : tool_rc_general_error;
        
        if (rc != tool_rc_success) {
            LOG_ERR("Could not load qualifier data");
            return rc;
        }
    }
    /*
     * 4. Configuration for calculating the pHash
     */

    /*
     * 4.a Determine pHash length and alg
     */
    ctx.cphash = ctx.cp_hash_path ? &ctx.cp_hash : 0;

    /*
     * 4.b Determine if TPM2_CC_<command> is to be dispatched
     */
    ctx.is_command_dispatch = ctx.cp_hash_path ? false : true;

    return rc;
}

static bool check_options(void) {

    if (ctx.cp_hash_path && (ctx.certify_info_path || ctx.signature_path)) {
        LOG_ERR("Cannot generate outputs when calculating cpHash.");
        return false;
    }

    if (!ctx.signing_key.ctx_path) {
        LOG_ERR("Must specify the signing key '-C'.");
        return false;
    }

    if (!ctx.certified_key.ctx_path) {
        LOG_ERR("Must specify the path of the key to certify '-c'.");
        return false;
    }

    if (!ctx.creation_ticket_path) {
        LOG_ERR("Must specify the creation ticket path '-t'.");
        return false;
    }

    if (!ctx.signature_path && !ctx.cp_hash_path) {
        LOG_ERR("Must specify the file path to save signature '-o'");
        return false;
    }

    if (!ctx.certify_info_path && !ctx.cp_hash_path) {
        LOG_ERR("Must specify file path to save attestation '--attestation'");
        return false;
    }

    return true;
}

static bool set_signature_format(char *value) {

    ctx.sig_format = tpm2_convert_sig_fmt_from_optarg(value);
    if (ctx.sig_format == signature_format_err) {
        return false;
    }
    return true;
}

static bool set_signing_scheme(char *value) {

    ctx.sig_scheme = tpm2_alg_util_from_optarg(value, tpm2_alg_util_flags_sig);
    if (ctx.sig_scheme == TPM2_ALG_ERROR) {
        LOG_ERR("Unknown signing scheme, got: \"%s\"", value);
        return false;
    }
    return true;
}

static bool set_digest_algorithm(char *value) {

    ctx.halg = tpm2_alg_util_from_optarg(value, tpm2_alg_util_flags_hash);
    if (ctx.halg == TPM2_ALG_ERROR) {
        LOG_ERR("Could not convert to number or lookup algorithm, got: "
                "\"%s\"", value);
        return false;
    }
    return true;
}

static bool on_option(char key, char *value) {

    bool result = true;

    switch (key) {
    case 'C':
        ctx.signing_key.ctx_path = value;
        break;
    case 'P':
        ctx.signing_key.auth_str = value;
        break;
    case 'c':
        ctx.certified_key.ctx_path = value;
        break;
    case 'd':
        ctx.creation_hash_path = value;
        break;
    case 't':
        ctx.creation_ticket_path = value;
        break;
    case 'g':
        result = set_digest_algorithm(value);
        goto on_option_out;
    case 's':
        result = set_signing_scheme(value);
        goto on_option_out;
    case 'f':
        result = set_signature_format(value);
        goto on_option_out;
    case 'o':
        ctx.signature_path = value;
        break;
    case 0:
        ctx.certify_info_path = value;
        break;
    case 1:
        ctx.cp_hash_path = value;
        break;
    case 'q':
        ctx.policy_qualifier_data = value;
        break;
        /* no default */
    }

on_option_out:
    return result;
}

static bool tpm2_tool_onstart(tpm2_options **opts) {

    static const struct option topts[] = {
      { "signingkey-context",   required_argument, NULL, 'C' },
      { "signingkey-auth",      required_argument, NULL, 'P' },
      { "certifiedkey-context", required_argument, NULL, 'c' },
      { "creation-hash",        required_argument, NULL, 'd' },
      { "ticket",               required_argument, NULL, 't' },
      { "hash-algorithm",       required_argument, NULL, 'g' },
      { "scheme",               required_argument, NULL, 's' },
      { "format",               required_argument, NULL, 'f' },
      { "signature",            required_argument, NULL, 'o' },
      { "attestation",          required_argument, NULL,  0  },
      { "qualification",        required_argument, NULL, 'q' },
      { "cphash",               required_argument, NULL,  1  },
    };

    *opts = tpm2_options_new("C:P:c:d:t:g:s:f:o:q:", ARRAY_LEN(topts), topts,
            on_option, NULL, 0);

    return *opts != NULL;
}

static tool_rc tpm2_tool_onrun(ESYS_CONTEXT *ectx, tpm2_option_flags flags) {

    UNUSED(flags);

    /*
     * 1. Process options
     */
    bool result = check_options();
    if (!result) {
        return tool_rc_option_error;
    }

   /*
     * 2. Process inputs
     */
    tool_rc rc = process_inputs(ectx);
    if (rc != tool_rc_success) {
        return rc;
    }

    /*
     * 3. TPM2_CC_<command> call
     */
    rc = certifycreation(ectx);
    if (rc != tool_rc_success) {
        return rc;
    }

    /*
     * 4. Process outputs
     */
    rc = process_output();

    return rc;
}

static tool_rc tpm2_tool_onstop(ESYS_CONTEXT *ectx) {

    UNUSED(ectx);

    /*
     * 1. Free objects
     */
    Esys_Free(ctx.signature);
    Esys_Free(ctx.certify_info);

    /*
     * 2. Close authorization sessions
     */
    return tpm2_session_close(&ctx.signing_key.object.session);

    /*
     * 3. Close auxiliary sessions
     */
}

// Register this tool with tpm2_tool.c
TPM2_TOOL_REGISTER("certifycreation", tpm2_tool_onstart, tpm2_tool_onrun,
tpm2_tool_onstop, NULL)
